modules/dnn/src/graph_simplifier.cpp

```diff
@@ -63,14 +63,23 @@ int Subgraph::getInputNodeId(const Ptr<ImportGraphWrapper>& net,
 {
     CV_Assert(inpId < node->getNumInputs());
     std::string name = node->getInputName(inpId);
+    std::string outName = name ;
     // If operation produces several tensors, they are specified by index
     // after ':' character. In example, "input:0".
-    name = name.substr(0, name.rfind(':'));
+    std:: string nodeName = name.substr(0, name.rfind(':'));
     const int numNodes = net->getNumNodes();
     for (int i = 0; i < numNodes; ++i)
     {
-        if (net->getNodeName(i) == name)
+        if (net->getNodeName(i) == nodeName)
```

Comment:
Node name can be not output name. So let's check only outputs names (let's TF graphs return `getNodeName` there)

Reply:
```
int Subgraph::getInputNodeId(const Ptr<ImportGraphWrapper>& net,
                             const Ptr<ImportNodeWrapper>& node,
                             int inpId)
{
    CV_Assert(inpId < node->getNumInputs());
    std::string name = node->getInputName(inpId);
    // If operation produces several tensors, they are specified by index
    // after ':' character. In example, "input:0".
    std::string name = name.substr(0, name.rfind(':'));
    const int numNodes = net->getNumNodes();
    for (int i = 0; i < numNodes; ++i)
    {
        if (net->getNodeName(i) == name)
            return i;
        Ptr<ImportNodeWrapper> node = net->getNode(i);
        const int numOutputs = node->getNumOutputs();
        for (int j = 0; j < numOutputs; j++)
        {
            if (node->getOutputName(j) == name)
                return i;
        }
    }
    CV_Error(Error::StsParseError, "Input node with name " + name + " not found");
}
```
@dkurt , is this okay ?

Reply:
I mean due we search input for some layer - it should be an output of another one, so I expected something like
```cpp
    for (int i = 0; i < numNodes; ++i)
    {
        Ptr<ImportNodeWrapper> node = net->getNode(i);
        const int numOutputs = node->getNumOutputs();
        for (int j = 0; j < numOutputs; j++)
        {
            if (node->getOutputName(j) == name)
                return i;
        }
    }
```

Reply:
for TF graphs, below statement will work
```
if (net->getNodeName(i) == name)
       return i;
```
for onnx graphs, we need to check both node name and all output names. Only checking output names wouldnt work. 

 ```
if (net->getNodeName(i) == name)
       return i;
Ptr<ImportNodeWrapper> node = net->getNode(i);
const int numOutputs = node->getNumOutputs();
for (int j = 0; j < numOutputs; j++)
{
      if (node->getOutputName(j) == name)
          return i;
}
```

Reply:
So for TF we can just override it
```cpp
int getNumOutputs() {
  return 1;
}

std::string getOutputName(int i) {
  return name;
}
```

Reply:
> I mean due we search input for some layer - it should be an output of another one, so I expected something like
> 
> ```c++
>     for (int i = 0; i < numNodes; ++i)
>     {
>         Ptr<ImportNodeWrapper> node = net->getNode(i);
>         const int numOutputs = node->getNumOutputs();
>         for (int j = 0; j < numOutputs; ++j)
>         {
>             if (node->getOutputName(j) == name)
>                 return i;
>         }
>     }
> ```

Okay, I got it. 
the problem with this codeis that sometimes some nodes have no outputs, like these nodes
![Screenshot from 2020-02-21 19-51-29](https://user-images.githubusercontent.com/26292819/75042176-04eae100-54e4-11ea-94bc-dec11b5c328c.png)

```
> number of nodes are 3 
> Node name : 0 number of outputs of node are : 0 having name 
> Node name : 1 number of outputs of node are : 0 having name 
> Node name : 2 number of outputs of node are : 1 having name 2 
```

there is no output of node 0 and 1. 
thats why we need to check both node and output names

Reply:
Yeah, but due it's ONNX specific problem - we need to keep it separately from common part. So for TF `getOutputName() === getNodeName` but for ONNX we can check
```cpp
getOutputName(int idx) {
  if (idx == 0)
    return name;
  else 
    output(idx)
}
```

Reply:
@dkurt 

```
int Subgraph::getInputNodeId(const Ptr<ImportGraphWrapper>& net,
                             const Ptr<ImportNodeWrapper>& node,
                             int inpId){
            ...
            Ptr<ImportNodeWrapper> node = net->getNode(i);
            const int numOutputs = node->getNumOutputs();
            ...
}
```

node is nullptr here, so it cannot be acessed using `class ImportNodeWrapper`. 

```
getOutputName(int idx) {
  if (idx == 0)
    return node->name;
  else 
   node->output(idx)
}
```
so defining getOutputName inside ImportNodeWrapper like this wouldnt work.
Node name can be acessed by `net->getNodeName(i)`

Reply:
**tf_graph_simplifier**
```
virtual int getNumOutputs() const CV_OVERRIDE
{
        return 1;
}
std::string getOutputName(int i) const CV_OVERRIDE
{
      return node->name();
}
```
**onnx_graph_simplifier**
```
virtual int getNumOutputs() const CV_OVERRIDE
{
    return node ? node->output_size() : 0;
}
virtual std::string getOutputName(int idx) const CV_OVERRIDE
{
    CV_Assert_N(node, idx < node->output_size());
     return node->output(idx);
}
```
**graph_simplifier**
```
int Subgraph::getInputNodeId(const Ptr<ImportGraphWrapper>& net,
                             const Ptr<ImportNodeWrapper>& node,
                             int inpId)
{
    CV_Assert(inpId < node->getNumInputs());
    std::string name = node->getInputName(inpId);
    // If operation produces several tensors, they are specified by index
    // after ':' character. In example, "input:0".
    name = name.substr(0, name.rfind(':'));
    const int numNodes = net->getNumNodes();
    for (int i = 0; i < numNodes; ++i)
    {
        Ptr<ImportNodeWrapper> node = net->getNode(i);
        int numOutputs = node->getNumOutputs();
        if (numOutputs == 0 && net->getNodeName(i) == name)
            return i;
        for (int j = 0; j < numOutputs; j++)
        {
            if (node->getOutputName(j) == name)
                return i;
        }
    }
    CV_Error(Error::StsParseError, "Input node with name " + name + " not found");
}
```
What about this implementation ?